### PR TITLE
CSS syntax module and error handling

### PR DESCRIPTION
**Pull request by estelle**
The original task was to update the charset module. Turns out there is no such thing.
Charset was defined in CSS 2 but removed in the CSS syntax module.

- Move css_charset to css_syntax.
- Created a new document: css_syntax. This is a module landing page. It includes the key concepts introduced in that module.
- Created a new document: a guide on CSS error handling by the parser. a subpage for CSS syntax.
- Updated the @charset page to indicate that it is NOT an at-rule, and has been deprecated in css-syntax.
- Updated custom-ident, the page that had the best explanation of CSS escape characters, adding a separate section for escaping characters in CSS. This PR does not include creating links to that section. That will be done as a separate PR (this PR is already big enough)
- related to https://github.com/openwebdocs/project/issues/186

This PR was previously reviewed and is now co-authored by @dipikabh. See https://github.com/mdn/content/pull/30793

### CODE REVIEW

File: files/en-us/web/css/css_syntax/error_handling/index.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+While invalid CSS is discarded with the property value falling back to the last valid value, an invalid computed custom property value is slightly different. If the custom property declaration is valid and the grammar of the property using that value is valid, the declarations are valid at computed time. The `var()` value substitution, however, can still be invalid.
+
+When a `var()` substitution is invalid, then the [initial](/en-US/docs/Web/CSS/initial_value) or [inherited](/en-US/docs/Web/CSS/Inheritance) value of the property is used; the declaration is not ignored. As the declaration was valid at compute time, the declaration is considered valid. This means the declaration is not ignored and the value of the previous valid value does not cascade. Rather, the property is set to a new value, but possibly not the expected one.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/css_syntax/error_handling/index.md:160]:
```suggestion
When a `var()` substitution is invalid, the declaration is not ignored and the [initial](/en-US/docs/Web/CSS/initial_value) or [inherited](/en-US/docs/Web/CSS/Inheritance) value of the property is used instead. The property is set to a new value, but possibly not the expected one.

Let's look at an example to illustrate this behavior:
```

File: files/en-us/web/css/custom-ident/index.md [hunk 2]
```diff
@@ -21,10 +21,27 @@ The syntax of `<custom-ident>` is similar to CSS identifiers (such as property n
 - any decimal digit (`0` to `9`),
 - a hyphen (`-`),
 - an underscore (`_`),
-- an escaped character (preceded by a backslash, `\`),
+- an [escaped character](#escaping-characters) (preceded by a backslash, `\`),
 - a [Unicode](https://en.wikipedia.org/wiki/Unicode) character (in the format of a backslash, `\`, followed by one to six hexadecimal digits, representing its Unicode code point)
 
-Note that `id1`, `Id1`, `iD1` and `ID1` are all different identifiers as they are [case-sensitive](https://en.wikipedia.org/wiki/Case_sensitivity). On the other hand, as there are several ways to escape a character, `toto\?` and `toto\3F` are the same identifiers.
+Note that `id1`, `Id1`, `iD1` and `ID1` are all different identifiers as they are [case-sensitive](https://en.wikipedia.org/wiki/Case_sensitivity).
```
Comment (chrisdavidmills) [inline files/en-us/web/css/custom-ident/index.md:27]:
```suggestion
Note that `id1`, `Id1`, `iD1`, and `ID1` are all different identifiers as they are [case-sensitive](https://en.wikipedia.org/wiki/Case_sensitivity).
```

File: files/en-us/web/css/custom-ident/index.md [hunk 3]
```diff
@@ -21,10 +21,27 @@ The syntax of `<custom-ident>` is similar to CSS identifiers (such as property n
...
+Any Unicode code point can be included as part of a `<custom-ident>` or quoted {{cssxref("string")}} by escaping it.
+
+In CSS, there are several ways to escape a character. Escape sequences start with a backslash (`\`), and continue with:
+
+- One to six hex (`ABCDEF0123456789`) digits. The hex digits can optionally be followed by white space. The hex escape sequence gets replaced by the Unicode code point whose value is given by those digits. The whitespace allows the sequences to be followed by actual hex digits (versus replaced ones).
+- Any Unicode code point that is not a hex digit or a newline character.
+
+Examples:
+
+- "&B" can be written as `\26 B` and `\000026B`.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/custom-ident/index.md:40]:
```suggestion
- "&B" can be written as `\26 B` or `\000026B`.
```

File: files/en-us/web/css/custom-ident/index.md [hunk 4]
```diff
@@ -21,10 +21,27 @@ The syntax of `<custom-ident>` is similar to CSS identifiers (such as property n
...
+- "hi.there" can be written as `hi\.there` and `hi\002Ethere`.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/custom-ident/index.md:41]:
```suggestion
- "hi.there" can be written as `hi\.there` or `hi\002Ethere`.
```

File: files/en-us/web/css/custom-ident/index.md [hunk 5]
```diff
@@ -21,10 +21,27 @@ The syntax of `<custom-ident>` is similar to CSS identifiers (such as property n
...
+- "toto?" can be written as`toto\?`, `toto\3F`, or `toto\00003F`
```
Comment (chrisdavidmills) [inline files/en-us/web/css/custom-ident/index.md:42]:
```suggestion
- "toto?" can be written as `toto\?`, `toto\3F`, or `toto\00003F`.
```

File: files/en-us/web/css/custom-ident/index.md [hunk 6]
```diff
@@ -21,10 +21,27 @@ The syntax of `<custom-ident>` is similar to CSS identifiers (such as property n
...
+
+To include actual white space after an escape sequence, include two white spaces in the escape sequence .
```
Comment (chrisdavidmills) [inline files/en-us/web/css/custom-ident/index.md:44]:
```suggestion
To include actual white space after an escape sequence, include two white spaces in the escape sequence.
```

Maybe include an example at the end of the sentence?

**Review by chrisdavidmills — CHANGES_REQUESTED** [review]
Hi @estelle! Thanks for a great new set of content. I've included quite a lot of comments, but these are mainly just quibbles about individual grammar and prose points. There is a lot of dense detail in the error handling page, and a lot of my suggestions are centered around trying to make this great content as clear as possible.

File: files/en-us/web/css/@charset/index.md [hunk 7]
```diff
@@ -7,21 +7,23 @@ browser-compat: css.at-rules.charset
 
 {{CSSRef}}
 
-The **`@charset`** [CSS](/en-US/docs/Web/CSS) [at-rule](/en-US/docs/Web/CSS/At-rule) specifies the character encoding used in the style sheet. It must be the first element in the style sheet and not be preceded by any character; as it is not a [nested statement](/en-US/docs/Web/CSS/Syntax#nested_statements), it cannot be used inside [conditional group at-rules](/en-US/docs/Web/CSS/At-rule#conditional_group_rules). If several `@charset` at-rules are defined, only the first one is used, and it cannot be used inside a `style` attribute on an HTML element or inside the {{ HTMLElement("style") }} element where the character set of the HTML page is relevant.
+The **`@charset`** [CSS](/en-US/docs/Web/CSS) rule specifies the character encoding used in the style sheet. Although the first character in `@charset` is the `@` symbol, it is not an [at-rule](/en-US/docs/Web/CSS/At-rule). The [CSS syntax](/en-US/docs/Web/CSS/CSS_syntax) module deprecates this rule, defining it as an unrecognized legacy rule to be dropped when a stylesheet is grammar-checked, but it is [fully supported in all browsers](#browser-compatibility).
```
Comment (estelle) [inline files/en-us/web/css/@charset/index.md:10]:
```suggestion
The **`@charset`** [CSS](/en-US/docs/Web/CSS) rule specifies the character encoding used in the style sheet. Although the first character in `@charset` is the `@` symbol, it is not an [at-rule](/en-US/docs/Web/CSS/At-rule). The [CSS syntax](/en-US/docs/Web/CSS/CSS_syntax) module deprecates this rule, defining it as an unrecognized legacy rule to be dropped when a stylesheet is grammar-checked, but it is [fully supported in all browsers](#browser_compatibility).
```

File: files/en-us/web/css/css_syntax/error_handling/index.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+If a selector list contains any invalid selectors, then the entire style block is ignored. There are exceptions: if the invalid selector is within the {{cssxref(":is")}} or {{cssxref(":where")}} pseudo-class, which accept [forgiving selector lists](/en-US/docs/Web/CSS/Selector_list#forgiving_selector_list) or if the unknown selector is a [`-webkit-` prefixed pseudo-element](#webkit-exception), only the unknown selector is ignored as not matching anything. The selector list is not invalidated.
+
+If a selector list contains any invalid selectors outside of these exceptions, a single invalid or unsupported selector in the selector list will invalidate the entire rule and the entire selector block will be ignored. The browser will look for the closing curly brace and will continue parsing again from that point on.
+
+#### Webkit exception
```
Comment (estelle) [inline files/en-us/web/css/css_syntax/error_handling/index.md:60]:
```suggestion
#### `-webkit-` exception
```

File: files/en-us/web/css/css_syntax/index.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+page-type: css-module
+spec-urls: https://drafts.csswg.org/css-syntax
+---
+
+{{CSSRef}}
+
+The **CSS syntax** module describes, in general terms, the structure and syntax of cascading stylesheets, or CSS. It defines CSS as the language for describing the rendering of structured documents (such as HTML and XML), on the web and elsewhere.
+
+This module doesn't define any properties, {{cssxref("css_types", "data types")}}, {{cssxref("css_functions", "functions")}}, or {{cssxref("at-rule", "at-rules")}}. Rather, it elaborates on how all of these features should be defined and how user agents should parse CSS.
+
```
Comment (estelle) [inline files/en-us/web/css/css_syntax/index.md:13]:
```suggestion
## At-rules

```

File: files/en-us/web/css/css_syntax/error_handling/index.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+The fact that browsers ignore invalid code enables the use of new CSS features without worrying about anything being broken in older browsers. A browser may not recognize a new feature, but that is okay. The discarding of invalid content without throwing an error also allows for both old and new syntaxes to coexist, in that order. Browsers render the old syntax until they recognize the new syntax as valid. Because browsers ignore invalid CSS, the valid fallbacks don't get overwritten by the new CSS that follows it that is perceived as invalid.
+
+The type and amount of CSS that a browser ignores due to an error depends on the type of error. Some common error situations are listed below:
+
+- For [errors in at-rules](#at-rule_errors), whether a single line or the entire at-rule is ignored (fails) depends on the at-rule and the type of error.
+- If the [error is an invalid selector](#errors_in_selector_lists), the entire declaration block is ignored.
+- An [error due to a missing semi-colon](#errors_within_css_declaration_blocks) between property declarations causes an invalid value, in which case, the property-value declaration is ignored.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_syntax/error_handling/index.md:27]: added the word "multiple". The rest is described in the linked section

File: files/en-us/web/css/css_syntax/error_handling/index.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+### Errors in selector lists
+
+There are many ways you might make a mistake writing a selector, but only invalid selectors cause a selector list to be invalid (See [invalid selector list](/en-US/docs/Web/CSS/Selector_list#invalid_selector_list)).
+
+If you include a {{cssxref("class_selectors", "class")}}, {{cssxref("id_selectors", "id")}}, or {{cssxref("type_selectors", "type")}} selector for a class, id, element or custom-element, respectively, that doesn't exist, it may be a logic error, but it's not a syntax error. If, however, you have a typo in a pseudo-class or a pseudo-element, it might create an invalid selector, which is an error the parser needs to address.
+
+If a selector list contains any invalid selectors, then the entire style block is ignored. There are exceptions: if the invalid selector is within the {{cssxref(":is")}} or {{cssxref(":where")}} pseudo-class, which accept [forgiving selector lists](/en-US/docs/Web/CSS/Selector_list#forgiving_selector_list) or if the unknown selector is a [`-webkit-` prefixed pseudo-element](#webkit-exception), only the unknown selector is ignored as not matching anything. The selector list is not invalidated.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_syntax/error_handling/index.md:60]: manually added

File: files/en-us/web/css/css_syntax/index.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+slug: Web/CSS/CSS_syntax
+page-type: css-module
+spec-urls: https://drafts.csswg.org/css-syntax
+---
+
+{{CSSRef}}
+
+The **CSS syntax** module describes, in general terms, the structure and syntax of cascading stylesheets, or CSS. It defines CSS as the language for describing the rendering of structured documents (such as HTML and XML), on the web and elsewhere.
+
+This module doesn't define any properties, {{cssxref("css_types", "data types")}}, {{cssxref("css_functions", "functions")}}, or {{cssxref("at-rule", "at-rules")}}. Rather, it elaborates on how all of these features should be defined and how user agents should parse CSS.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_syntax/index.md:12]: thanks. I did not know this trick.

File: files/en-us/web/css/css_syntax/index.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+[WHATWG](/en-US/docs/Glossary/WHATWG) specification:
+
+- {{HTMLElement("style")}} element
+- {{HTMLElement("link")}} element
+- [`class`](/en-US/docs/Web/HTML/Global_attributes/class) attribute
+- [`rel`](/en-US/docs/Web/HTML/Attributes/rel#stylesheet) attribute
+
+## Specifications
+
+{{Specifications}}
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_syntax/index.md:100]: i know. working on it. But that is a different repo in a different org.

**Review by estelle — COMMENTED** [review]
@chrisdavidmills Thanks for the feedback. I think i captured everything.

File: files/en-us/web/css/css_syntax/error_handling/index.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+You may come across legacy CSS that looks like the following:
+
+```css example-bad
+/* Prefixed values */
+.wrapper {
+  display: -webkit-box;
+  display: -webkit-flex;
+  display: -ms-flexbox;
+  display: flex;
```
Comment (estelle) [inline files/en-us/web/css/css_syntax/error_handling/index.md:108]:
```suggestion
  display: flex;
  display: block flex;
```

File: files/en-us/web/css/css_syntax/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/css/css_syntax/index.md]: Oops, looks like the links got broken. As I feared, that trick doesn't work in cssxref, only domxref. I've fixed the links for you as a penance for wasting your time.

File: files/en-us/web/css/css_syntax/error_handling/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/css/css_syntax/error_handling/index.md]: Yup, I think that works.

**Review by chrisdavidmills — APPROVED** [review]
Looking awesome, nice work @estelle (Simmons). Let's get this merged.

**Pull request merged**